### PR TITLE
SQL Server Migrations: Don't rebuild indexes which will be created

### LINKED ISSUE #7535: SQL Server Migrations: AlterColumn tries to drop non-existing index
Hi,

I have the exact same problem as described in https://github.com/aspnet/EntityFramework/issues/7396, when upgrading from "IdentityServer4.EntityFramework": "1.0.0-rc1-update1" to "IdentityServer4.EntityFramework": "1.0.0". I already have a database built on the RC version, and I am now trying to upgrade to version 1.0.0. This is a context defined by the IdentityServer team, so I am just doing the upgrade. I want to do the migrations as part of my upgrade, ideally as part of the server startup. 

The problem I face lies in PersistedGrantDbContext. The snapshot from the RC version is:
```
    [DbContext(typeof(PersistedGrantDbContext))]
    partial class PersistedGrantDbContextModelSnapshot : ModelSnapshot
    {
        protected override void BuildModel(ModelBuilder modelBuilder)
        {
            modelBuilder
                .HasAnnotation("ProductVersion", "1.0.0-rtm-21431")
                .HasAnnotation("SqlServer:ValueGenerationStrategy", SqlServerValueGenerationStrategy.IdentityColumn);

            modelBuilder.Entity("IdentityServer4.EntityFramework.Entities.PersistedGrant", b =>
                {
                    b.Property<string>("Key");
                    b.Property<string>("Type");
                    b.Property<string>("ClientId")
                        .IsRequired()
                        .HasAnnotation("MaxLength", 200);
                    b.Property<DateTime>("CreationTime");
                    b.Property<string>("Data")
                        .IsRequired();
                    b.Property<DateTime>("Expiration");
                    b.Property<string>("SubjectId");
                    b.HasKey("Key", "Type");
                    b.ToTable("PersistedGrants");
                });
        }
    }
```
Please note that there is no HasIndex here. I then upgrade to version 1.0.0, and create a migration. The generated migration is then:
```
public partial class PersistedGrantDbUpgradeTo100Migration : Migration
    {
        protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.AlterColumn<string>(
                name: "SubjectId",
                table: "PersistedGrants",
                maxLength: 200,
                nullable: true);

            migrationBuilder.AlterColumn<DateTime>(
                name: "Expiration",
                table: "PersistedGrants",
                nullable: true);

            migrationBuilder.AlterColumn<string>(
                name: "Type",
                table: "PersistedGrants",
                maxLength: 50,
                nullable: false);

            migrationBuilder.AlterColumn<string>(
                name: "Key",
                table: "PersistedGrants",
                maxLength: 200,
                nullable: false);

            migrationBuilder.CreateIndex(
                name: "IX_PersistedGrants_SubjectId",
                table: "PersistedGrants",
                column: "SubjectId");

            migrationBuilder.CreateIndex(
                name: "IX_PersistedGrants_SubjectId_ClientId",
                table: "PersistedGrants",
                columns: new[] { "SubjectId", "ClientId" });

            migrationBuilder.CreateIndex(
                name: "IX_PersistedGrants_SubjectId_ClientId_Type",
                table: "PersistedGrants",
                columns: new[] { "SubjectId", "ClientId", "Type" });
        }
...
```
When starting the server, I call the following:
```
		private void initializeDatabase(IApplicationBuilder app)
		{
			using (var scope = app.ApplicationServices.GetService<IServiceScopeFactory>().CreateScope())
			{
				scope.ServiceProvider.GetRequiredService<PersistedGrantDbContext>().Database.Migrate();
...
```
This then creates the following upgrade script:
```
DROP INDEX [IX_PersistedGrants_SubjectId] ON [PersistedGrants];
DROP INDEX [IX_PersistedGrants_SubjectId_ClientId] ON [PersistedGrants];
DROP INDEX [IX_PersistedGrants_SubjectId_ClientId_Type] ON [PersistedGrants];
DECLARE @var0 sysname;
SELECT @var0 = [d].[name]
FROM [sys].[default_constraints] [d]
INNER JOIN [sys].[columns] [c] ON [d].[parent_column_id] = [c].[column_id] AND [d].[parent_object_id] = [c].[object_id]
WHERE ([d].[parent_object_id] = OBJECT_ID(N'PersistedGrants') AND [c].[name] = N'SubjectId');
IF @var0 IS NOT NULL EXEC(N'ALTER TABLE [PersistedGrants] DROP CONSTRAINT [' + @var0 + '];');
ALTER TABLE [PersistedGrants] ALTER COLUMN [SubjectId] nvarchar(200);
CREATE INDEX [IX_PersistedGrants_SubjectId] ON [PersistedGrants] ([SubjectId]);
CREATE INDEX [IX_PersistedGrants_SubjectId_ClientId] ON [PersistedGrants] ([SubjectId], [ClientId]);
CREATE INDEX [IX_PersistedGrants_SubjectId_ClientId_Type] ON [PersistedGrants] ([SubjectId], [ClientId], [Type]);
```
And then the exception: 
```
Unhandled Exception: System.Data.SqlClient.SqlException: Cannot drop the index 'PersistedGrants.IX_PersistedGrants_SubjectId', because it does not exist or you do not have permission.
Cannot drop the index 'PersistedGrants.IX_PersistedGrants_SubjectId_ClientId', because it does not exist or you do not have permission.
Cannot drop the index 'PersistedGrants.IX_PersistedGrants_SubjectId_ClientId_Type', because it does not exist or you do not have permission.
Warning! The maximum key length is 900 bytes. The index 'IX_PersistedGrants_SubjectId_ClientId_Type' has maximum length of 1700 bytes. For some combination of large values, the insert/update operation will fail.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlCommand.RunExecuteNonQueryTds(String methodName, Boolean async, Int32 timeout, Boolean asyncWrite)
   at System.Data.SqlClient.SqlCommand.InternalExecuteNonQuery(TaskCompletionSource`1 completion, Boolean sendToPipe, Int32 timeout, Boolean asyncWrite, String methodName)
   at System.Data.SqlClient.SqlCommand.ExecuteNonQuery()
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, String executeMethod, IReadOnlyDictionary`2 parameterValues, Boolean closeConnection)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteNonQuery(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationCommandExecutor.ExecuteNonQuery(IEnumerable`1 migrationCommands, IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Migrate(String targetMigration)
```
### Further technical details
EF Core version: 1.1.0
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10 64 bit
IDE: Visual Studio 2015 Update 3

Best regards,
Ivan Uthus

### PR DESCRIPTION
**Pull request by bricelam**
Fixes #7535

### CODE REVIEW

File: src/Microsoft.EntityFrameworkCore.SqlServer/Migrations/Internal/SqlServerMigrationsSqlGenerator.cs [hunk 1]
```diff
@@ -967,19 +986,67 @@ protected override void ForeignKeyAction(ReferentialAction referentialAction, Mi
...
+                    if (beforeOperation)
+                    {
+                        if (operation == currentOperation)
+                        {
+                            beforeOperation = false;
+                            continue;
+                        }
+
+                        if (operation is DropIndexOperation dropOperation
+                            && dropOperation.Name == indexName)
```
Comment (bricelam) [inline src/Microsoft.EntityFrameworkCore.SqlServer/Migrations/Internal/SqlServerMigrationsSqlGenerator.cs]: Hmm, this shouldn't actually be needed... I'll remove and bash some more.

Comment (bricelam) [inline src/Microsoft.EntityFrameworkCore.SqlServer/Migrations/Internal/SqlServerMigrationsSqlGenerator.cs]: Confirmed. You'd have to be doing some contradictory manual edits to hit this case. Not worth handling.

**Comment by ajcvickers** [comment]
:shipit:

**Comment by ErikEJ** [comment]
providers-fyi ?

**Comment by bricelam** [comment]
Nope, just @ErikEJ-fyi since it's SQL Server-specific. 😉

**Comment by ErikEJ** [comment]
LOL 😆 - thanks!

**Pull request merged**